From 952c3169b6037a284d36fedd4971f955454bca81 Mon Sep 17 00:00:00 2001
From: Wang Zaikuo <zaikuo.wang@intel.com>
Date: Mon, 7 May 2018 12:38:21 +0800
Subject: [PATCH 1892/2313] media: intel-ipu4: isys: tpg support
 store_csi2_header

Message for Open Source:
Currently, mipi_store_mode value in stream_cfg is 1 for TPG pipe
by default; but TPG subdev did not have store_csi2_header support;
This causes that user has no chance to disable
mipi_store_mode and captured image size is larger than expection

This patch add support for store_csi2_header in TPG subdev and
set enabled by default; if user captures on tpg video device,
it could set ctrl value 0 to disable mipi header output like this:
	yavta -w "0x00981982 0" $dev_node

Message for Internal:
[Issue/Feature]
refer to open source message

[Root Cause/Changes]
refer to open source message

Change-Id: Iad8bf46184e244997911806993936bfa8e0a5adb
Signed-off-by: Wang Zaikuo <zaikuo.wang@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 drivers/media/pci/intel/ipu-isys-tpg.c   | 32 +++++++++++++++++++++++++++++++-
 drivers/media/pci/intel/ipu-isys-tpg.h   |  1 +
 drivers/media/pci/intel/ipu-isys-video.c |  4 ++++
 drivers/media/pci/intel/ipu-isys-video.h |  2 --
 4 files changed, 36 insertions(+), 3 deletions(-)

diff --git a/drivers/media/pci/intel/ipu-isys-tpg.c b/drivers/media/pci/intel/ipu-isys-tpg.c
index 5b8b940..53ab79e 100644
--- a/drivers/media/pci/intel/ipu-isys-tpg.c
+++ b/drivers/media/pci/intel/ipu-isys-tpg.c
@@ -105,6 +105,16 @@ static struct v4l2_ctrl_config tpg_mode = {
 	.qmenu = tpg_mode_items,
 };
 
+static const struct v4l2_ctrl_config csi2_header_cfg = {
+	.id = V4L2_CID_IPU_STORE_CSI2_HEADER,
+	.name = "Store CSI-2 Headers",
+	.type = V4L2_CTRL_TYPE_BOOLEAN,
+	.min = 0,
+	.max = 1,
+	.step = 1,
+	.def = 1,
+};
+
 static void ipu_isys_tpg_init_controls(struct v4l2_subdev *sd)
 {
 	struct ipu_isys_tpg *tpg = to_ipu_isys_tpg(sd);
@@ -150,6 +160,8 @@ static void ipu_isys_tpg_init_controls(struct v4l2_subdev *sd)
 	}
 
 	v4l2_ctrl_new_custom(&tpg->asd.ctrl_handler, &tpg_mode, NULL);
+	tpg->store_csi2_header =
+		v4l2_ctrl_new_custom(&tpg->asd.ctrl_handler, &csi2_header_cfg, NULL);
 }
 
 static void tpg_set_ffmt(struct v4l2_subdev *sd,
@@ -197,7 +209,25 @@ const struct ipu_isys_pixelformat *ipu_isys_tpg_try_fmt(struct ipu_isys_video
 							v4l2_pix_format_mplane
 							*mpix)
 {
-	return ipu_isys_video_try_fmt_vid_mplane(av, mpix, 1);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 5, 0)
+	struct media_entity entity = av->vdev.entity;
+	struct v4l2_subdev *sd =
+		media_entity_to_v4l2_subdev(entity.links[0].source->entity);
+#else
+	struct media_link *link = list_first_entry(&av->vdev.entity.links,
+						   struct media_link, list);
+	struct v4l2_subdev *sd =
+		media_entity_to_v4l2_subdev(link->source->entity);
+#endif
+	struct ipu_isys_tpg *tpg;
+
+	if (!sd)
+		return NULL;
+
+	tpg = to_ipu_isys_tpg(sd);
+
+	return ipu_isys_video_try_fmt_vid_mplane(av, mpix,
+		v4l2_ctrl_g_ctrl(tpg->store_csi2_header));
 }
 
 static const struct v4l2_subdev_pad_ops tpg_sd_pad_ops = {
diff --git a/drivers/media/pci/intel/ipu-isys-tpg.h b/drivers/media/pci/intel/ipu-isys-tpg.h
index d782e1d..aae1ca6 100644
--- a/drivers/media/pci/intel/ipu-isys-tpg.h
+++ b/drivers/media/pci/intel/ipu-isys-tpg.h
@@ -81,6 +81,7 @@ struct ipu_isys_tpg {
 	struct v4l2_ctrl *llp;
 	struct v4l2_ctrl *fll;
 	struct v4l2_ctrl *pixel_rate;
+	struct v4l2_ctrl *store_csi2_header;
 };
 
 #define to_ipu_isys_tpg(sd)		\
diff --git a/drivers/media/pci/intel/ipu-isys-video.c b/drivers/media/pci/intel/ipu-isys-video.c
index a3af92c..dac46e8 100644
--- a/drivers/media/pci/intel/ipu-isys-video.c
+++ b/drivers/media/pci/intel/ipu-isys-video.c
@@ -1059,6 +1059,9 @@ static int start_stream_firmware(struct ipu_isys_video *av,
 	if (ip->csi2 && !v4l2_ctrl_g_ctrl(ip->csi2->store_csi2_header))
 		stream_cfg->input_pins[0].mipi_store_mode =
 		    IPU_FW_ISYS_MIPI_STORE_MODE_DISCARD_LONG_HEADER;
+	else if (ip->tpg && !v4l2_ctrl_g_ctrl(ip->tpg->store_csi2_header))
+		stream_cfg->input_pins[0].mipi_store_mode =
+		    IPU_FW_ISYS_MIPI_STORE_MODE_DISCARD_LONG_HEADER;
 
 	stream_cfg->src = ip->source;
 	stream_cfg->vc = 0;
@@ -1387,6 +1390,7 @@ int ipu_isys_video_prepare_streaming(struct ipu_isys_video *av,
 	ip->csi2_be = NULL;
 	ip->csi2_be_soc = NULL;
 	ip->csi2 = NULL;
+	ip->tpg = NULL;
 	ip->seq_index = 0;
 	memset(ip->seq, 0, sizeof(ip->seq));
 
diff --git a/drivers/media/pci/intel/ipu-isys-video.h b/drivers/media/pci/intel/ipu-isys-video.h
index 9b7da62..c7a264b 100644
--- a/drivers/media/pci/intel/ipu-isys-video.h
+++ b/drivers/media/pci/intel/ipu-isys-video.h
@@ -54,9 +54,7 @@ struct ipu_isys_pipeline {
 	struct ipu_isys_csi2_be *csi2_be;
 	struct ipu_isys_csi2_be_soc *csi2_be_soc;
 	struct ipu_isys_csi2 *csi2;
-#ifdef IPU_TPG_SOF
 	struct ipu_isys_tpg *tpg;
-#endif
 	/*
 	 * Number of capture queues, write access serialised using struct
 	 * ipu_isys.stream_mutex
-- 
2.7.4

